File: crocoddyl/kkt.py
import numpy as np


class SolverKKT:
    def __init__(self,shootingProblem):
        self.problem = shootingProblem
    
        self.nx  = sum([ m.nx  for m in self.models() ])
        self.ndx = sum([ m.ndx for m in self.models() ])
        self.nu  = sum([ m.nu  for m in self.problem.runningModels ])
        
        self.kkt    = np.zeros([ 2*self.ndx+self.nu, 2*self.ndx+self.nu ])
        self.kktref = np.zeros(  2*self.ndx+self.nu )

        self.hess   = self.kkt[:self.ndx+self.nu,:self.ndx+self.nu]
        self.jac    = self.kkt[self.ndx+self.nu:,:self.ndx+self.nu]
        self.jacT   = self.kkt[:self.ndx+self.nu,self.ndx+self.nu:]
        self.grad   = self.kktref[:self.ndx+self.nu]
        self.cval   = self.kktref[self.ndx+self.nu:]

        self.Lxx    = self.hess[:self.ndx,:self.ndx]
        self.Lxu    = self.hess[:self.ndx,self.ndx:]
        self.Lux    = self.hess[self.ndx:,:self.ndx]
        self.Luu    = self.hess[self.ndx:,self.ndx:]
        self.Lx     = self.grad[:self.ndx]
        self.Lu     = self.grad[self.ndx:]
        self.Fx     = self.jac[:,:self.ndx]
        self.Fu     = self.jac[:,self.ndx:]

        self.alphas = [ 10**(-n) for n in range(7) ]
        self.th_acceptStep = .1
        self.th_stop = 1e-9
        self.th_grad = 1e-12

        self.x_reg = 0
        self.u_reg = 0
        
    def models(self) : return self.problem.runningModels + [self.problem.terminalModel]
    def datas(self) : return self.problem.runningDatas + [self.problem.terminalData]
    def setCandidate(self,xs=None,us=None,isFeasible=False,copy=True):
        '''
        Set the solver candidate value for the decision variables, as a trajectory xs,us
        of T+1 and T elements. isFeasible should be set to True if the xs are 
        obtained from integrating the us (roll-out).
        If copy is True, make a copy of the data.
        '''
        if xs is None: xs = [ m.State.zero() for m in self.models() ]
        elif copy:     xs = [ x.copy() for x in xs ]
        if us is None: us = [ np.zeros(m.nu) for m in self.problem.runningModels ]
        elif copy:     us = [ u.copy() for u in us ]

        assert( len(xs) == self.problem.T+1 )
        assert( len(us) == self.problem.T )
        self.xs = xs
        self.us = us
        self.isFeasible = isFeasible

    def calc(self):
        '''
        For a given pair of candidate state and control trajectories,
        compute the KKT system. 
        isFeasible should be set True if F(xs,us) = 0 (xs obtained from rollout).
        '''
        xs,us = self.xs,self.us
        problem = self.problem
        self.cost = problem.calcDiff(xs,us)
        ix = 0; iu = 0
        cx0 = problem.runningModels[0].ndx  # offset on constraint xnext = f(x,u) due to x0 = ref.

        np.fill_diagonal(self.Fx,1)
        for model,data,xguess in zip(problem.runningModels,problem.runningDatas,xs[1:]):
            ndx = model.ndx; nu = model.nu
            self.Lxx[ix:ix+ndx,ix:ix+ndx] = data.Lxx
            self.Lxu[ix:ix+ndx,iu:iu+nu ] = data.Lxu
            self.Lux[iu:iu+nu ,ix:ix+ndx] = data.Lxu.T
            self.Luu[iu:iu+nu ,iu:iu+nu ] = data.Luu

            self.Lx[ix:ix+ndx]           = data.Lx
            self.Lu[iu:iu+nu ]           = data.Lu
            
            self.Fx[cx0+ix:cx0+ix+ndx,ix:ix+ndx] = - data.Fx
            self.Fu[cx0+ix:cx0+ix+ndx,iu:iu+nu ] = - data.Fu

            # constraint value = xnext_guess - f(x_guess,u_guess) = diff(f,xnext_guesss)
            self.cval[cx0+ix:cx0+ix+ndx] = model.State.diff(data.xnext,xguess) #data.F
            ix += ndx; iu += nu

        model,data = problem.terminalModel,problem.terminalData
        ndx = model.ndx; nu = model.nu
        self.Lxx[ix:ix+ndx,ix:ix+ndx] = data.Lxx
        self.Lx [ix:ix+ndx]           = data.Lx

        # constraint value = x_guess - x_ref = diff(x_ref,x_guess)
        self.cval[:cx0] = problem.runningModels[0].State.diff(problem.initialState,xs[0])

        self.jacT[:] = self.jac.T

        # Regularization
        ndx = sum([ m.ndx for m in self.models() ])
        nu  = sum([ m.nu  for m in self.problem.runningModels ])
        if self.x_reg != 0: self.kkt[range(ndx),       range(ndx)       ] += self.x_reg
        if self.u_reg != 0: self.kkt[range(ndx,ndx+nu),range(ndx,ndx+nu)] += self.u_reg
                
        return self.cost

    def computePrimalDual(self):
        self.primaldual = np.linalg.solve(self.kkt,-self.kktref)
        self.primal = self.primaldual[:self.ndx+self.nu]
        self.dual = self.primaldual[-self.ndx:]
        
    def computeDirection(self,recalc=True):
        '''
        Compute the direction of descent for the current guess xs,us. 
        if recalc is True, run self.calc() before hand. self.setCandidate 
        must have been called before.
        '''
        if recalc: self.calc()
        self.computePrimalDual()

        p_x  = self.primaldual[:self.ndx]
        p_u  = self.primaldual[self.ndx:self.ndx+self.nu]
        ix = 0; iu = 0
        dxs = []; dus = []; lambdas = []
        for model,data in zip(self.problem.runningModels,self.problem.runningDatas):
            ndx = model.ndx; nu = model.nu
            dxs.append( p_x[ix:ix+ndx] )
            dus.append( p_u[iu:iu+nu] )
            lambdas.append( self.dual[ix:ix+ndx] )
            ix+=ndx; iu+=nu
        dxs.append( p_x[-self.problem.terminalModel.ndx:] )
        lambdas.append( self.dual[-self.problem.terminalModel.ndx:] )
        self.dxs = dxs ; self.dus = dus ; self.lambdas = lambdas
        return dxs,dus,lambdas

    def expectedImprovement(self):
        '''
        Return the expected improvement that the current step should bring,
        i.e dV_exp = f(xk) - f_exp(xk + delta), where f_exp(xk+delta) = f_k + m_k(delta)
        and m_k is the quadratic model of f at x_k, and delta is the descent direction.
        Then: dv_exp = - grad*delta - .5*hess*delta**2
        '''
        return -np.dot(self.grad,self.primal), \
            -np.dot(np.dot(self.hess,self.primal),self.primal)

    def stoppingCriteria(self):
        '''
        Return two terms:
        * sumsq(dL+dF) = || d/dx Cost - d/dx (lambda^T constraint) ||^2
        * sumsq(cval)  = || d/dlambda L ||^2 = || constraint ||^2
        '''
        lambdas = self.lambdas
        dL = self.grad
        dF = np.concatenate([ lk - np.dot(lkp1,dk.Fx)
                              for lk,dk,lkp1 in zip( lambdas[:-1],
                                                     self.problem.runningDatas,
                                                     lambdas[1:]) ] \
                            + [ lambdas[-1] ] \
                            + [ -np.dot(lkp1,dk.Fu)
                                for lk,dk,lkp1 in zip( lambdas[:-1],
                                                       self.problem.runningDatas,
                                                       lambdas[1:]) ])
        return sum((dL+dF)**2), sum(self.cval**2)
        
    def tryStep(self,stepLength):
        '''
        Compute in x_try,u_try a step by adding stepLength*dx to x.
        Store the result in self.xs_try,self.us_try.
        Return the cost improvement ie cost(xs,us)-cost(xs_try,us_try).
        '''
        self.xs_try = [ m.State.integrate(x,stepLength*dx)
                        for m,x,dx in zip(self.models(),self.xs,self.dxs) ]
        self.us_try = [ u+stepLength*du for u,du in zip(self.us,self.dus) ]
        self.cost_try = self.problem.calc(self.xs_try,self.us_try)
        dV  = self.cost-self.cost_try
        return dV
    
    def solve(self,maxiter=100,init_xs = None,init_us = None,isFeasible=False,verbose=False):
        '''
        From an initial guess init_xs,init_us (feasible or not),
        iterate over computeDirection and tryStep until stoppingCriteria is below threshold.
        '''
        self.setCandidate(init_xs,init_us,isFeasible,copy=True)
        for i in range(maxiter):
            self.computeDirection()
            d1,d2 = self.expectedImprovement()

            for a in self.alphas:
                dV = self.tryStep(a)
                if verbose: print('\t\tAccept? %f %f' % (dV, d1*a+.5*d2*a**2) )
                if d1<1e-9 or not isFeasible or dV > self.th_acceptStep*(d1*a+.5*d2*a**2):
                    # Accept step
                    self.setCandidate(self.xs_try,self.us_try,isFeasible=True,copy=False)
                    break
            if verbose: print( 'Accept iter=%d, a=%f, cost=%.8f'
                               % (i,a,self.problem.calc(self.xs,self.us)))
                
            self.stop = sum(self.stoppingCriteria())
            if self.stop<self.th_stop:
                return self.xs,self.us,True
            # if d1<self.th_grad:
            #     return self.xs,self.us,False

        # Warning: no convergence in max iterations
        return self.xs,self.us,False

File: src/core/solvers/kkt.cpp
///////////////////////////////////////////////////////////////////////////////
// BSD 3-Clause License
//
// Copyright (C) 2018-2020, LAAS-CNRS, New York University, Max Planck Gesellschaft,
//                          University of Edinburgh
// Copyright note valid unless otherwise stated in individual files.
// All rights reserved.
///////////////////////////////////////////////////////////////////////////////

#include "crocoddyl/core/solvers/kkt.hpp"

namespace crocoddyl {

SolverKKT::SolverKKT(boost::shared_ptr<ShootingProblem> problem)
    : SolverAbstract(problem),
      regfactor_(10.),
      regmin_(1e-9),
      regmax_(1e9),
      cost_try_(0.),
      th_grad_(1e-12),
      th_step_(0.5),
      was_feasible_(false) {
  allocateData();
  const unsigned int& n_alphas = 10;
  xreg_ = 0.;
  ureg_ = 0.;
  alphas_.resize(n_alphas);
  for (unsigned int n = 0; n < n_alphas; ++n) {
    alphas_[n] = 1. / pow(2., (double)n);
  }
}

SolverKKT::~SolverKKT() {}

bool SolverKKT::solve(const std::vector<Eigen::VectorXd>& init_xs, const std::vector<Eigen::VectorXd>& init_us,
                      const std::size_t& maxiter, const bool& is_feasible, const double&) {
  setCandidate(init_xs, init_us, is_feasible);
  bool recalc = true;
  for (iter_ = 0; iter_ < maxiter; ++iter_) {
    while (true) {
      try {
        computeDirection(recalc);
      } catch (const char* msg) {
        recalc = false;
        if (xreg_ == regmax_) {
          return false;
        } else {
          continue;
        }
      }
      break;
    }

    expectedImprovement();
    for (std::vector<double>::const_iterator it = alphas_.begin(); it != alphas_.end(); ++it) {
      steplength_ = *it;
      try {
        dV_ = tryStep(steplength_);
      } catch (const char* msg) {
        continue;
      }
      dVexp_ = steplength_ * d_[0] + 0.5 * steplength_ * steplength_ * d_[1];
      if (d_[0] < th_grad_ || !is_feasible_ || dV_ > th_acceptstep_ * dVexp_) {
        was_feasible_ = is_feasible_;
        setCandidate(xs_try_, us_try_, true);
        cost_ = cost_try_;
        break;
      }
    }
    stoppingCriteria();
    const std::size_t& n_callbacks = callbacks_.size();
    if (n_callbacks != 0) {
      for (std::size_t c = 0; c < n_callbacks; ++c) {
        CallbackAbstract& callback = *callbacks_[c];
        callback(*this);
      }
    }
    if (was_feasible_ && stop_ < th_stop_) {
      return true;
    }
  }
  return false;
}

void SolverKKT::computeDirection(const bool& recalc) {
  const std::size_t& T = problem_->get_T();
  if (recalc) {
    calc();
  }
  computePrimalDual();
  const Eigen::VectorBlock<Eigen::VectorXd, Eigen::Dynamic> p_x = primal_.segment(0, ndx_);
  const Eigen::VectorBlock<Eigen::VectorXd, Eigen::Dynamic> p_u = primal_.segment(ndx_, nu_);

  std::size_t ix = 0;
  std::size_t iu = 0;
  for (std::size_t t = 0; t < T; ++t) {
    const std::size_t& ndxi = problem_->running_models_[t]->get_state()->get_ndx();
    const std::size_t& nui = problem_->running_models_[t]->get_nu();
    dxs_[t] = p_x.segment(ix, ndxi);
    dus_[t] = p_u.segment(iu, nui);
    lambdas_[t] = dual_.segment(ix, ndxi);
    ix += ndxi;
    iu += nui;
  }
  const std::size_t& ndxi = problem_->terminal_model_->get_state()->get_ndx();
  dxs_.back() = p_x.segment(ix, ndxi);
  lambdas_.back() = dual_.segment(ix, ndxi);
}

double SolverKKT::tryStep(const double& steplength) {
  const std::size_t& T = problem_->get_T();
  for (std::size_t t = 0; t < T; ++t) {
    const boost::shared_ptr<ActionModelAbstract>& m = problem_->running_models_[t];

    m->get_state()->integrate(xs_[t], steplength * dxs_[t], xs_try_[t]);
    us_try_[t] = us_[t];
    us_try_[t] += steplength * dus_[t];
  }
  const boost::shared_ptr<ActionModelAbstract> m = problem_->terminal_model_;
  m->get_state()->integrate(xs_[T], steplength * dxs_[T], xs_try_[T]);
  cost_try_ = problem_->calc(xs_try_, us_try_);
  return cost_ - cost_try_;
}

double SolverKKT::stoppingCriteria() {
  const std::size_t& T = problem_->get_T();
  std::size_t ix = 0;
  std::size_t iu = 0;
  for (std::size_t t = 0; t < T; ++t) {
    const boost::shared_ptr<ActionDataAbstract>& d = problem_->running_datas_[t];
    const std::size_t& ndxi = problem_->running_models_[t]->get_state()->get_ndx();
    const std::size_t& nui = problem_->running_models_[t]->get_nu();

    dF.segment(ix, ndxi) = lambdas_[t];
    dF.segment(ix, ndxi).noalias() -= d->Fx.transpose() * lambdas_[t + 1];
    dF.segment(ndx_ + iu, nui).noalias() = -lambdas_[t + 1].transpose() * d->Fu;
    ix += ndxi;
    iu += nui;
  }
  const std::size_t& ndxi = problem_->terminal_model_->get_state()->get_ndx();
  dF.segment(ix, ndxi) = lambdas_.back();
  stop_ = (kktref_.segment(0, ndx_ + nu_) + dF).squaredNorm() + kktref_.segment(ndx_ + nu_, ndx_).squaredNorm();
  return stop_;
}

const Eigen::Vector2d& SolverKKT::expectedImprovement() {
  d_ = Eigen::Vector2d::Zero();
  // -grad^T.primal
  d_(0) = -kktref_.segment(0, ndx_ + nu_).dot(primal_);
  // -(hessian.primal)^T.primal
  kkt_primal_.noalias() = kkt_.block(0, 0, ndx_ + nu_, ndx_ + nu_) * primal_;
  d_(1) = -kkt_primal_.dot(primal_);
  return d_;
}

const Eigen::MatrixXd& SolverKKT::get_kkt() const { return kkt_; }

const Eigen::VectorXd& SolverKKT::get_kktref() const { return kktref_; }

const Eigen::VectorXd& SolverKKT::get_primaldual() const { return primaldual_; }

const std::vector<Eigen::VectorXd>& SolverKKT::get_dxs() const { return dxs_; }

const std::vector<Eigen::VectorXd>& SolverKKT::get_dus() const { return dus_; }

const std::size_t& SolverKKT::get_nx() const { return nx_; }

const std::size_t& SolverKKT::get_ndx() const { return ndx_; }

const std::size_t& SolverKKT::get_nu() const { return nu_; }

double SolverKKT::calc() {
  cost_ = problem_->calcDiff(xs_, us_);

  // offset on constraint xnext = f(x,u) due to x0 = ref.
  const std::size_t& cx0 = problem_->get_runningModels()[0]->get_state()->get_ndx();

  std::size_t ix = 0;
  std::size_t iu = 0;
  const std::size_t& T = problem_->get_T();
  kkt_.block(ndx_ + nu_, 0, ndx_, ndx_) = Eigen::MatrixXd::Identity(ndx_, ndx_);
  for (std::size_t t = 0; t < T; ++t) {
    const boost::shared_ptr<ActionModelAbstract>& m = problem_->running_models_[t];
    const boost::shared_ptr<ActionDataAbstract>& d = problem_->running_datas_[t];
    const std::size_t& ndxi = m->get_state()->get_ndx();
    const std::size_t& nui = m->get_nu();

    // Computing the gap at the initial state
    if (t == 0) {
      m->get_state()->diff(problem_->get_x0(), xs_[0], kktref_.segment(ndx_ + nu_, ndxi));
    }

    // Filling KKT matrix
    kkt_.block(ix, ix, ndxi, ndxi) = d->Lxx;
    kkt_.block(ix, ndx_ + iu, ndxi, nui) = d->Lxu;
    kkt_.block(ndx_ + iu, ix, nui, ndxi) = d->Lxu.transpose();
    kkt_.block(ndx_ + iu, ndx_ + iu, nui, nui) = d->Luu;
    kkt_.block(ndx_ + nu_ + cx0 + ix, ix, ndxi, ndxi) = -d->Fx;
    kkt_.block(ndx_ + nu_ + cx0 + ix, ndx_ + iu, ndxi, nui) = -d->Fu;

    // Filling KKT vector
    kktref_.segment(ix, ndxi) = d->Lx;
    kktref_.segment(ndx_ + iu, nui) = d->Lu;
    m->get_state()->diff(d->xnext, xs_[t + 1], kktref_.segment(ndx_ + nu_ + cx0 + ix, ndxi));

    ix += ndxi;
    iu += nui;
  }
  boost::shared_ptr<ActionDataAbstract>& df = problem_->terminal_data_;
  const std::size_t& ndxf = problem_->terminal_model_->get_state()->get_ndx();
  kkt_.block(ix, ix, ndxf, ndxf) = df->Lxx;
  kktref_.segment(ix, ndxf) = df->Lx;
  kkt_.block(0, ndx_ + nu_, ndx_ + nu_, ndx_) = kkt_.block(ndx_ + nu_, 0, ndx_, ndx_ + nu_).transpose();
  return cost_;
}

void SolverKKT::computePrimalDual() {
  primaldual_ = kkt_.lu().solve(-kktref_);
  primal_ = primaldual_.segment(0, ndx_ + nu_);
  dual_ = primaldual_.segment(ndx_ + nu_, ndx_);
}

void SolverKKT::increaseRegularization() {
  xreg_ *= regfactor_;
  if (xreg_ > regmax_) {
    xreg_ = regmax_;
  }
  ureg_ = xreg_;
}

void SolverKKT::decreaseRegularization() {
  xreg_ /= regfactor_;
  if (xreg_ < regmin_) {
    xreg_ = regmin_;
  }
  ureg_ = xreg_;
}

void SolverKKT::allocateData() {
  const std::size_t& T = problem_->get_T();
  dxs_.resize(T + 1);
  dus_.resize(T);
  lambdas_.resize(T + 1);
  xs_try_.resize(T + 1);
  us_try_.resize(T);

  nx_ = 0;
  ndx_ = 0;
  nu_ = 0;
  for (std::size_t t = 0; t < T; ++t) {
    const boost::shared_ptr<ActionModelAbstract>& model = problem_->running_models_[t];
    const std::size_t& nx = model->get_state()->get_nx();
    const std::size_t& ndx = model->get_state()->get_ndx();
    const std::size_t& nu = model->get_nu();
    if (t == 0) {
      xs_try_[t] = problem_->get_x0();
    } else {
      xs_try_[t] = Eigen::VectorXd::Constant(nx, NAN);
    }
    us_try_[t] = Eigen::VectorXd::Constant(nu, NAN);
    dxs_[t] = Eigen::VectorXd::Zero(ndx);
    dus_[t] = Eigen::VectorXd::Zero(nu);
    lambdas_[t] = Eigen::VectorXd::Zero(ndx);
    nx_ += nx;
    ndx_ += ndx;
    nu_ += nu;
  }
  const boost::shared_ptr<ActionModelAbstract>& model = problem_->get_terminalModel();
  nx_ += model->get_state()->get_nx();
  ndx_ += model->get_state()->get_ndx();
  xs_try_.back() = problem_->terminal_model_->get_state()->zero();
  dxs_.back() = Eigen::VectorXd::Zero(model->get_state()->get_ndx());
  lambdas_.back() = Eigen::VectorXd::Zero(model->get_state()->get_ndx());

  // Set dimensions for kkt matrix and kkt_ref vector
  kkt_.resize(2 * ndx_ + nu_, 2 * ndx_ + nu_);
  kkt_.setZero();
  kktref_.resize(2 * ndx_ + nu_);
  kktref_.setZero();
  primaldual_.resize(2 * ndx_ + nu_);
  primaldual_.setZero();
  primal_.resize(ndx_ + nu_);
  primal_.setZero();
  kkt_primal_.resize(ndx_ + nu_);
  kkt_primal_.setZero();
  dual_.resize(ndx_);
  dual_.setZero();
  dF.resize(ndx_ + nu_);
  dF.setZero();
}

}  // namespace crocoddyl
